src/target/nrf54l.c

```diff
@@ -0,0 +1,189 @@
+#define NRF54L_CTRL_AP_APPROTECT_STATUS_SECUREAPPROTECT_ENABLED (1U << 1U)
+
+static bool rram_erase(target_flash_s *flash, target_addr_t addr, size_t len)
+{
+	(void)flash;
+	(void)addr;
+	(void)len;
+	// RRAM doesn't need to be erased before being written, so we just return ok.
+	return true;
+}
+
+static bool rram_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t len)
+{
+	while (target_mem32_read32(flash->t, NRF54L_RRAMC_BUFSTATUS_READYNEXT) != 1U) {
+		// Wait for rram to be ready for next write.
+	}
+	target_mem32_write(flash->t, dest, src, len);
+	return true;
+}
+
+static bool rram_prepare(target_flash_s *flash)
+{
+	uint32_t writebufsize = flash->writesize / 16U;
+	target_mem32_write32(
+		flash->t, NRF54L_RRAMC_CONFIG, NRF54L_RRAMC_CONFIG_WRITEBUFSIZE(writebufsize) | NRF54L_RRAMC_CONFIG_WEN);
+	return true;
+}
+
+static bool rram_done(target_flash_s *flash)
+{
+	while (target_mem32_read32(flash->t, NRF54L_RRAMC_BUFSTATUS_WRITEBUFEMPTY) != 1U) {
+		// Wait for writebuf to flush.
+	}
+	target_mem32_write32(flash->t, NRF54L_RRAMC_CONFIG, 0U);
+	return true;
+}
+
+static bool rram_mass_erase(target_s *const target, platform_timeout_s *const print_progess)
+{
+	target_mem32_write32(target, NRF54L_RRAMC_ERASE_ERASEALL, 1U);
+
+	uint32_t ready;
+	do {
+		ready = target_mem32_read32(target, NRF54L_RRAMC_READY);
+		target_print_progress(print_progess);
+	} while (ready == 0U);
+
+	return true;
+}
+
+static void add_rram(target_s *target, uint32_t addr, size_t length, uint32_t writesize)
+{
+	target_flash_s *flash = calloc(1, sizeof(*flash));
+	if (!flash) { /* calloc failed: heap exhaustion */
+		DEBUG_ERROR("calloc: failed in %s\n", __func__);
+		return;
+	}
+
+	flash->start = addr;
+	flash->length = length;
+	flash->blocksize = writesize;
+	flash->writesize = writesize;
+	flash->erase = rram_erase;
+	flash->write = rram_write;
+	flash->prepare = rram_prepare;
+	flash->done = rram_done;
+	flash->erased = 0xffU;
+	target_add_flash(target, flash);
+}
+
+bool nrf54l_probe(target_s *target)
+{
+	adiv5_access_port_s *ap = cortex_ap(target);
+
+	if (ap->dp->version < 2U)
+		return false;
+
+	switch (ap->dp->target_partno) {
+	case NRF54L_PARTNO:
+		target->driver = "Nordic nRF54L";
+		target->target_options |= TOPT_INHIBIT_NRST;
+		break;
+	default:
+		return false;
+	}
+
+	target->mass_erase = rram_mass_erase;
+
+	uint32_t info_ram = target_mem32_read32(target, NRF54L_FICR_INFO_RAM);
+	uint32_t info_rram = target_mem32_read32(target, NRF54L_FICR_INFO_RRAM);
+
+	target_add_ram32(target, NRF54L_RAM, info_ram * 1024U);
+	add_rram(target, NRF54L_RRAM, info_rram * 1024U, 512U);
+	add_rram(target, NRF54L_UICR, 0x1000U, 4U);
+
+	return true;
+}
+
+static bool nrf54l_ctrl_ap_mass_erase(target_s *target, platform_timeout_s *print_progess);
+
+bool nrf54l_ctrl_ap_probe(adiv5_access_port_s *ap)
+{
+	switch (ap->idr) {
+	case NRF54L_CTRL_AP_IDR_VALUE:
+		break;
+	default:
+		return false;
+	}
+
+	target_s *t = target_new();
```

Comment:
It seems this one got missed in all the fixes for single letter variable names. Please use `target` here.